src/TaskLineRenderer.ts

```diff
@@ -0,0 +1,228 @@
+import { Component, MarkdownRenderer } from 'obsidian';
+import type { Moment } from 'moment';
+import type { Task } from './Task';
+import * as taskModule from './Task';
+import type { LayoutComponent, LayoutOptions } from './TaskLayout';
+import { TaskLayout } from './TaskLayout';
+import { replaceTaskWithTasks } from './File';
+import { getSettings } from './Config/Settings';
+
+export type TaskLineRenderDetails = {
+    parentUlElement: HTMLElement;
+    /** The nth item in this list (including non-tasks). */
+    listIndex: number;
+    layoutOptions?: LayoutOptions;
+    isFilenameUnique?: boolean;
+};
+
+/**
+ * The function used to render a Markdown task line into an existing HTML element.
+ */
+export type TextRenderer = (text: string, element: HTMLSpanElement, path: string) => Promise<void>;
+
+async function obsidianMarkdownRenderer(text: string, element: HTMLSpanElement, path: string) {
+    await MarkdownRenderer.renderMarkdown(text, element, path, null as unknown as Component);
+}
+
+/**
+ * Renders a given Task object into an HTML List Item (LI) element, using the given renderDetails
+ * configuration and a supplied TextRenderer (typically the Obsidian Markdown renderer, but for testing
+ * purposes it can be a simpler one).
+ * The element includes the task and its various components (description, priority, block link etc), the
+ * checkbox on the left with its event handling of completing the task, and the button for editing the task.
+ */
+export async function renderTaskLine(
+    task: Task,
+    renderDetails: TaskLineRenderDetails,
+    textRenderer: TextRenderer | null = null,
+): Promise<HTMLLIElement> {
+    if (!textRenderer) textRenderer = obsidianMarkdownRenderer;
+    const li: HTMLLIElement = document.createElement('li');
+    renderDetails.parentUlElement.appendChild(li);
+    li.classList.add('task-list-item', 'plugin-tasks-list-item');
+
+    const textSpan = document.createElement('span');
+    li.appendChild(textSpan);
+    textSpan.classList.add('tasks-list-text');
+    await taskToHtml(task, renderDetails, textSpan, textRenderer);
+
+    const checkbox = document.createElement('input');
+    li.appendChild(checkbox);
+    checkbox.classList.add('task-list-item-checkbox');
+    checkbox.type = 'checkbox';
+    if (task.status !== taskModule.Status.TODO) {
+        checkbox.checked = true;
+        li.classList.add('is-checked');
+    }
```

Comment:
A small point, but 44 to 56 are a bit dense to read. A comment or two might help clarify the structure..

I think a part of it may be the pattern of moving from lines like this:

```ts
const checkbox = li.createEl('input');
```

to lines like this:

```ts
    const checkbox = document.createElement('input');
    li.appendChild(checkbox);
```

For my own education, what's the benefit of the new version?

Reply:
Will add a few comments to explain the flow.
Regarding the change in how the various elements are created, `li.createEl` is a shortcut provided for convenience by the Obsidian API, but we don't have it available when running tests, and I wanted the tests to be able to create the full div and span structure, so had to convert all of these to the equivalent but more elaborate `document.createElement` and `appendChild` calls.

Reply:
Actually there's more to that which is worth sharing. I'm dreaming of creating an Android home screen query widget that can do stuff like presenting today's tasks on the phone home screen etc, so in addition of solving a practical problem, I thought it's worthwhile to decrease the dependency of the rendering code on Obsidian's API.
Given the limitations of the free time I'm able to find this will likely stay in the dream scope, but making it a step more reachable in case someone else wants to do it looked worthwhile :)

Reply:
Thanks for both those helpful answers!

Reply:
Thanks again. I've added a comment in the code explaining why the more verbose form is used.